### ISSUE TITLE
How to Subscribe for events

### ISSUE THREAD

**Issue by ddobric** [dapr/dotnet-sdk#288]
Looking for an example which demonstrates how to subscribe events.
The provided Client sample shows how to publish an event, but there is not way in the DaprClient to receive  events.

Especially looking for PUB/SUB concepts like retries, peek-and-lock, peek-and-delete,..
This is all requirement on the concrete component, but do not see anything in the spec related to such concepts.
PUB/SUB building block is definitely extremely important , but common brokers support many integration patterns and features related to them. My question is how to deal with them in dapr.
Publishing an event and receiving t is very basic operation, which will not satisfy most of common enterprise requirements.

Thanks in advance

**Comment by amanbha** [comment]
For subscribing to events please see, https://github.com/dapr/docs/blob/master/reference/api/pubsub_api.md#required-application-user-code-routes

If you are using aspnetcore, then you can use the aspnetcore samples to subscribe to events which adds a route for the topics as well.

Dapr building block provides abstraction over multiple components so full fledged functionality is not available, however components allow extending this functionality by providing metadata on requests/response.

For specific scenarios in pub/sub building block, please open issue in dapr runtime repo or discuss it in our gitter channel https://gitter.im/Dapr/community

**Comment by ddobric** [comment]
Thanks for your feedback @amanbha .
I guess we are missing a method in DaprClientSdk to subscribe for events or similar. 

For example:

`var eventReader = client.SubscribeAsync() `

**Comment by amanbha** [comment]
Subscribed events are discovered by Dapr by calling on to `http://localhost:<appPort>/dapr/subscribe` endpoint to discover subscribed topics. DaprClientSDK provides methods to call into Dapr and doesnt start a web server, so route has to be configured in user app. If Aspnet core integration with dapr is used then topics can be subscribed as shown in the Routing sample and the aspnetcore integration will configure the correct routes. https://github.com/dapr/dotnet-sdk/tree/master/samples/AspNetCore/RoutingSample

**Comment by tommorvolloriddle** [comment]
Hello @amanbha 
I have a similar problem at receiving end, I have is a c++ monolith application and needs dapr for pub-sub, if I am unable to run dapr side-by-side to c++  application (as the application bootstraps with some scripts) and have no way for declaring call backs (with HTTP verb attributed methods) will I be still able to subscribe to messages, Just for your information: I am using RabbitMQ for messaging

**Comment by tommorvolloriddle** [comment]
> Subscribed events are discovered by Dapr by calling on to `http://localhost:<appPort>/dapr/subscribe` endpoint to discover subscribed topics. DaprClientSDK provides methods to call into Dapr and doesnt start a web server, so route has to be configured in user app. If Aspnet core integration with dapr is used then topics can be subscribed as shown in the Routing sample and the aspnetcore integration will configure the correct routes. https://github.com/dapr/dotnet-sdk/tree/master/samples/AspNetCore/RoutingSample

Hello @amanbha 
How will I subscribe to events without using aspnet core application, I have a class library from which I need to subscribe to dapr events, I am using rabbitmq as message broker and I am running dapr to which I will connect my application

**Comment by WhitWaldo** [comment]
@tommorvolloriddle Going off your comment in June, are you using the C# SDK or the C++ SDK? If the former, you're in the right place, but if the latter, you need to pose this question in the [C++ SDK repository](https://github.com/dapr/cpp-sdk/).

If you're using the C# SDK as mentioned earlier, the C# SDK does not spin up a web server on your behalf, so you need something that can handle inbound GET and POST requests to URLs. Typically in the C# universe, that's going to be ASP.NET Core and if you're using that, we've got several helper extensions and attributes you can use to decorate your endpoints accordingly. 

If you're using some other C# web framework, I'd appreciate it if you'd tell me which one so we might consider it for prioritization and you'll have to handle mapping inbound requests per the [PubSub API specification](https://docs.dapr.io/reference/api/pubsub_api/).

**Comment by tommorvolloriddle** [comment]
> @tommorvolloriddle Going off your comment in June, are you using the C# SDK or the C++ SDK? If the former, you're in the right place, but if the latter, you need to pose this question in the [C++ SDK repository](https://github.com/dapr/cpp-sdk/).
> 
> If you're using the C# SDK as mentioned earlier, the C# SDK does not spin up a web server on your behalf, so you need something that can handle inbound GET and POST requests to URLs. Typically in the C# universe, that's going to be ASP.NET Core and if you're using that, we've got several helper extensions and attributes you can use to decorate your endpoints accordingly.
> 
> If you're using some other C# web framework, I'd appreciate it if you'd tell me which one so we might consider it for prioritization and you'll have to handle mapping inbound requests per the [PubSub API specification](https://docs.dapr.io/reference/api/pubsub_api/).

@WhitWaldo 
I am using C# SDK , Things were bit hazy in June, but now its clear that I will use C# SDK, however the usecase is this, I have a C# client library, which gets integrated in another application that cannot be side by side to dapr (using dapr run and dotnet run combination) so I am creating client from the class library to achieve publish and subscribe workflow

Publish workflow seems to be fairly straight forward I am dong something like this:

``
  public async Task PublishMessageAsync<T>(string topic, T message) where T : IMessage
  {
      myDaprClient = new DaprClientBuilder().UseGrpcEndpoint("http://localhost:50001/v1.0/publish/").Build();
      var content = message.ToByteArray();
      await myDaprClient.PublishEventAsync("rabbits", topic, content);       

  }
``

But on the subscriber part I didn't get Subscribe API hence I attempted to create HttpClientListner object and listen on Dapr subscribe endpoint something like this 

``
static async Task Main(string[] args)
{
    string prefix = "http://localhost:50001/dapr/subscribe/"; // Adjust the endpoint as needed
    _httpListener.Prefixes.Add(prefix);
    _httpListener.Start();
    Console.WriteLine($"Listening for messages on {prefix}");

    while (true)
    {
        var context = await _httpListener.GetContextAsync();
        _ = ProcessMessageAsync(context);
    }
}
``
This didn't work very well and also dynamically binding messages to handlers based on the topics that they are arriving was not clear to me, to sum it up there is no success story to me on this approach.

I reached to the git repo of Dapr.Messaging to realize that there is support to Subscibe via SubscibeAsync with Streaming subscriptions.

With SubscribeAsync I am now able to create client to dapr sidecar and subscribe to events by doming something like this:

``
 internal void SubscribeToTopicAsync(string topic)
 {
     myDaprPublishSubscribeClient = new DaprPublishSubscribeClientBuilder().UseGrpcEndpoint("http://localhost:50001/v1.0/subscribe").Build();
     Task.Run(async () =>
     {
         await myDaprPublishSubscribeClient.SubscribeAsync("rabbits", topic,
               new DaprSubscriptionOptions(new MessageHandlingPolicy(Timeout.InfiniteTimeSpan, TopicResponseAction.Drop)),
               messageHandler: Handler);
     });

 }
``
In a nutshell I want the publish and subscribe to happen in C# client library by creating clients to dapr side car running in standalone mode
